File: fixtures/projects/e2e-app/src/payments/processor.ts
export type PaymentRequest = {
	amount: number;
	currency: string;
	reference: string;
	metadata: Record<string, string>;
};

export type PaymentReceipt = {
	id: string;
	provider: string;
	approved: boolean;
	amount: number;
	currency: string;
};

export interface PaymentProcessor {
	readonly provider: string;
	processPayment(request: PaymentRequest): Promise<PaymentReceipt>;
	refundPayment(receipt: PaymentReceipt): Promise<boolean>;
}

export abstract class PaymentProcessor {
	abstract readonly provider: string;

	protected normalizeAmount(amount: number): number {
		return validateInput(amount);
	}

	abstract processPayment(request: PaymentRequest): Promise<PaymentReceipt>;
	abstract refundPayment(receipt: PaymentReceipt): Promise<boolean>;
}

export function validateInput(amount: number): number {
	if (!Number.isFinite(amount) || amount <= 0) {
		throw new Error("Payment amount must be a positive number");
	}
	return Math.round(amount * 100) / 100;
}

File: fixtures/projects/e2e-python/src/payments/processor.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Protocol


class PaymentAuditHook(Protocol):
    def record(self, provider: str, reference: str) -> None: ...


@dataclass(slots=True)
class PaymentRequest:
    amount: float
    currency: str
    reference: str


@dataclass(slots=True)
class PaymentReceipt:
    provider: str
    reference: str
    approved: bool


class PaymentProcessor(ABC):
    provider: str

    @abstractmethod
    def process_payment(self, request: PaymentRequest) -> PaymentReceipt:
        raise NotImplementedError


def process_payment(
    processor: PaymentProcessor,
    request: PaymentRequest,
    audit: PaymentAuditHook | None = None,
) -> PaymentReceipt:
    receipt = processor.process_payment(request)
    if audit is not None:
        audit.record(receipt.provider, receipt.reference)
    return receipt
